File: sosyolobi-web-2/src/lib/date-filters.ts
export type DateFilterKey = "all" | "today" | "week" | "month";

export const DATE_FILTER_OPTIONS: { label: string; value: DateFilterKey }[] = [
  { label: "Tümü", value: "all" },
  { label: "Bugün", value: "today" },
  { label: "Bu Hafta", value: "week" },
  { label: "Bu Ay", value: "month" },
];

function startOfDay(d: Date) {
  const r = new Date(d);
  r.setHours(0, 0, 0, 0);
  return r;
}

function endOfDay(d: Date) {
  const r = new Date(d);
  r.setHours(23, 59, 59, 999);
  return r;
}

/** Pazartesi başlangıçlı hafta */
function startOfWeek(d: Date) {
  const r = startOfDay(d);
  const day = (r.getDay() + 6) % 7; // Pzt=0 ... Paz=6
  r.setDate(r.getDate() - day);
  return r;
}

function endOfWeek(d: Date) {
  const r = startOfWeek(d);
  r.setDate(r.getDate() + 6);
  return endOfDay(r);
}

function startOfMonth(d: Date) {
  return startOfDay(new Date(d.getFullYear(), d.getMonth(), 1));
}

function endOfMonth(d: Date) {
  return endOfDay(new Date(d.getFullYear(), d.getMonth() + 1, 0));
}

export function getDateRange(key: DateFilterKey, now: Date = new Date()): { fromDate?: string; toDate?: string } {
  switch (key) {
    case "today":
      return { fromDate: startOfDay(now).toISOString(), toDate: endOfDay(now).toISOString() };
    case "week":
      return { fromDate: startOfWeek(now).toISOString(), toDate: endOfWeek(now).toISOString() };
    case "month":
      return { fromDate: startOfMonth(now).toISOString(), toDate: endOfMonth(now).toISOString() };
    default:
      return {};
  }
}

export function isWeekendDate(iso: string): boolean {
  const day = new Date(iso).getDay();
  return day === 0 || day === 6; // Pazar veya Cumartesi
}

File: sosyolobi-mobile/lib/core/utils/date_filters.dart
/// Ports `sosyolobi-web-2/src/lib/date-filters.ts` verbatim — client-side
/// date-range + weekend-only filtering for the activities list and map
/// screens. Backend already supports `fromDate`/`toDate` query params on
/// `/api/activities/nearby` and `/api/activities/map` (no backend change
/// needed); weekend filtering is purely client-side (`eventDate`'s local
/// weekday is Saturday/Sunday).
enum DateFilterKey { all, today, week, month }

const dateFilterOptions = [
  (label: 'Tümü', value: DateFilterKey.all),
  (label: 'Bugün', value: DateFilterKey.today),
  (label: 'Bu Hafta', value: DateFilterKey.week),
  (label: 'Bu Ay', value: DateFilterKey.month),
];

abstract final class DateFilters {
  static DateTime _startOfDay(DateTime d) => DateTime(d.year, d.month, d.day);

  static DateTime _endOfDay(DateTime d) => DateTime(d.year, d.month, d.day, 23, 59, 59, 999);

  /// Monday-start week, mirroring web's `startOfWeek`.
  static DateTime _startOfWeek(DateTime d) {
    final start = _startOfDay(d);
    final daysFromMonday = start.weekday - DateTime.monday; // Mon=0 ... Sun=6
    return start.subtract(Duration(days: daysFromMonday));
  }

  static DateTime _endOfWeek(DateTime d) => _endOfDay(_startOfWeek(d).add(const Duration(days: 6)));

  static DateTime _startOfMonth(DateTime d) => DateTime(d.year, d.month, 1);

  static DateTime _endOfMonth(DateTime d) => _endOfDay(DateTime(d.year, d.month + 1, 0));

  /// Returns the local-time `(fromDate, toDate)` bounds for [key] — callers
  /// should send `.toUtc()` when passing these as query params, matching the
  /// backend's `.ToUniversalTime()` normalization.
  static ({DateTime? fromDate, DateTime? toDate}) rangeFor(DateFilterKey key, [DateTime? now]) {
    final n = now ?? DateTime.now();
    switch (key) {
      case DateFilterKey.today:
        return (fromDate: _startOfDay(n), toDate: _endOfDay(n));
      case DateFilterKey.week:
        return (fromDate: _startOfWeek(n), toDate: _endOfWeek(n));
      case DateFilterKey.month:
        return (fromDate: _startOfMonth(n), toDate: _endOfMonth(n));
      case DateFilterKey.all:
        return (fromDate: null, toDate: null);
    }
  }

  static bool isWeekend(DateTime eventDate) {
    final local = eventDate.toLocal();
    return local.weekday == DateTime.saturday || local.weekday == DateTime.sunday;
  }
}
